pkg/asset/installconfig/aws/validation.go

```diff
@@ -166,6 +173,126 @@ func validateAMI(ctx context.Context, config *types.InstallConfig) field.ErrorLi
+		return nil
+	} else if len(config.Platform.AWS.EIPAllocations.IngressNetworkLoadBalancer) > 0 {
+		eipAllocations = config.Platform.AWS.EIPAllocations.IngressNetworkLoadBalancer
+	}
+	var allocationIDs []string
+	for _, allocationID := range eipAllocations {
+		allocationIDs = append(allocationIDs, string(allocationID))
+	}
+
+	sess, err := meta.Session(ctx)
+	if err != nil {
+		return append(allErrs, field.Invalid(fldPath, nil, fmt.Sprintf("unable to start a session: %s", err.Error())))
+	}
+
+	client := ec2.New(sess, aws.NewConfig().WithRegion(config.Platform.AWS.Region))
+
+	// Check if eipAllocations exist
+	for _, allocationID := range allocationIDs {
+		_, err := eipAllocationExists(client, allocationID)
+		if err != nil {
+			err = fmt.Errorf("eipAllocations does not exists due to error %s:", err)
+			return append(allErrs, field.InternalError(fldPath, err))
+		}
+	}
+
+	// Get all associated EIPs in the list
+	associatedEIPs, err := getAssociatedEIPs(client, allocationIDs)
+	if err != nil {
+		err = fmt.Errorf("could not get check if eips are associated due to error %s:", err)
+		return append(allErrs, field.InternalError(fldPath, err))
+	}
+
+	// Print error if EIPs are associated
+	if len(associatedEIPs) > 0 {
+		return append(allErrs, field.Invalid(fldPath, associatedEIPs, fmt.Errorf("eipAllocations %s can't be used as %s are already associated", config.Platform.AWS.EIPAllocations.IngressNetworkLoadBalancer, associatedEIPs).Error()))
+	}
+
+	// BYO Subnets
+	if len(config.Platform.AWS.Subnets) != 0 {
+		if err = compareBYOSubnetsWithEIPAllocations(config); err != nil {
+			return append(allErrs, field.Invalid(fldPath, nil, compareBYOSubnetsWithEIPAllocations(config).Error()))
+		}
+	} else { // Installer Created Subnets
+		allzones, err := meta.AvailabilityZones(ctx)
+		if err != nil {
+			return append(allErrs, field.InternalError(fldPath, err))
+		}
+		if err = compareInstallerCreatedSubnetsWithEIPAllocations(config, allzones); err != nil {
+			return append(allErrs, field.Invalid(fldPath, nil, compareInstallerCreatedSubnetsWithEIPAllocations(config, allzones).Error()))
+		}
+	}
+	return nil
+}
+
+func compareBYOSubnetsWithEIPAllocations(config *types.InstallConfig) error {
+	if len(config.Platform.AWS.Subnets) != len(config.Platform.AWS.EIPAllocations.IngressNetworkLoadBalancer) {
+		return fmt.Errorf("subnet count %s != eipAllocations count %s", config.Platform.AWS.Subnets, config.Platform.AWS.EIPAllocations.IngressNetworkLoadBalancer)
+	}
+	return nil
+}
+
+func compareInstallerCreatedSubnetsWithEIPAllocations(config *types.InstallConfig, allzones []string) error {
+	// Compare eip allocations with Availability Zones.
+	if config.Platform.AWS != nil && config.Platform.AWS.EIPAllocations != nil && len(config.Platform.AWS.EIPAllocations.IngressNetworkLoadBalancer) != 0 && len(allzones) != len(config.Platform.AWS.EIPAllocations.IngressNetworkLoadBalancer) {
```

Comment:
it would make this more readable by only passing `config.Platform.AWS` to the function (rather than `config`) and replacing all instances of `config.Platform.AWS` with aws